gem/refactorise.py

```diff
@@ -112,6 +116,135 @@ def product(*args):
             result.add(sum_indices, atomics, rest)
         return result
 
+    def argument_indices_extent(self, factor):
+        if self.argument_indices is None:
+            raise AssertionError("argument_indices property not initialised.")
+        return numpy.product([i.extent for i in set(factor.free_indices).intersection(self.argument_indices)])
+
+    def all_sum_indices(self):
+        result = []
+        collected = set()
+        for (sum_indices_set, _), (sum_indices, _) in iteritems(self.ordering):
+            if sum_indices_set not in collected:
+                result.append((sum_indices_set, sum_indices))
+                collected.add(sum_indices_set)
+        return result
+
+    def to_expression(self):
+        # No argument factorisation here yet
+        indexsums = []
+        for sum_indices_set, sum_indices in self.all_sum_indices():
+            all_atomics = []
+            all_rest = []
+            for key, (_, atomics) in iteritems(self.ordering):
+                if sum_indices_set == key[0]:
```

Comment:
This pattern where you agglomerate by unique indices by looping over all the unique indices and then looping over all the non-unique indices, checking whether you match the one you currently want appears multiple times.  Is it not rather written something like this, where you first agglomerate, and then work on the agglomerated things.

```python
all_atomics = OrderedDict()
indexsums = []
for key, (_, atomics) in iteritems(self.ordering):
    rest = self.monomials[key]
    sum_indices, _ = key
    if not sum_indices:
        indexsums.append(fast_sum_factorise((), atomics + (rest, )))
    else:
        all_atomics.setdefault(sum_indices, []).append((atomics, rest))
for sum_indices, (atomics, rest) in iteritems(all_atomics):
    if len(atomics) == 1:
        indexsums.append(fast_sum_factorise(sum_indices, zip(atomics, rest)))
    else:
        products = [associate_product(atom + (rest_, ))[0] for atom, rest_ in zip(atomics, rest))]
        indexsums.append(...)
```
untested.